dozer-api/src/grpc/mod.rs

```diff
@@ -11,23 +11,143 @@ mod shared_impl;
 
+use crate::shutdown::ShutdownReceiver;
+
+#[derive(Debug)]
+enum ShutdownAddrStream<F> {
+    Alive { inner: AddrStream, shutdown: F },
+    Shutdown { inner: AddrStream },
+    Temp,
+}
+
+impl<F: Future<Output = ()> + Unpin> ShutdownAddrStream<F> {
+    fn check_shutdown(&mut self, cx: &mut Context<'_>) -> Result<(), io::Error> {
+        loop {
+            match self {
+                Self::Alive { shutdown, .. } => {
+                    if let Poll::Ready(()) = Pin::new(shutdown).poll(cx) {
+                        let mut temp = Self::Temp;
+                        std::mem::swap(self, &mut temp);
+                        let Self::Alive { inner, .. } = temp else {
+                            unreachable!()
+                        };
+                        *self = Self::Shutdown { inner };
+                        continue;
+                    } else {
+                        return Ok(());
+                    }
+                }
+                Self::Shutdown { inner } => {
+                    if let Poll::Ready(Err(e)) = Pin::new(inner).poll_shutdown(cx) {
+                        return Err(e);
+                    } else {
+                        return Ok(());
+                    }
+                }
+                Self::Temp => unreachable!(),
+            }
+        }
+    }
+
+    fn pool_impl<T>(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        func: fn(Pin<&mut AddrStream>, &mut Context<'_>) -> Poll<io::Result<T>>,
+    ) -> Poll<io::Result<T>> {
+        let this = Pin::into_inner(self);
+        if let Err(e) = this.check_shutdown(cx) {
+            return Poll::Ready(Err(e));
+        }
+
+        match this {
+            Self::Alive { inner, .. } => func(Pin::new(inner), cx),
+            Self::Shutdown { inner } => func(Pin::new(inner), cx),
+            Self::Temp => unreachable!(),
+        }
+    }
+}
+
+impl<F: Future<Output = ()> + Unpin> AsyncRead for ShutdownAddrStream<F> {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        let this = Pin::into_inner(self);
+        if let Err(e) = this.check_shutdown(cx) {
+            return Poll::Ready(Err(e));
+        }
+
+        match this {
+            Self::Alive { inner, .. } => Pin::new(inner).poll_read(cx, buf),
+            Self::Shutdown { inner } => Pin::new(inner).poll_read(cx, buf),
+            Self::Temp => unreachable!(),
+        }
+    }
+}
+
+impl<F: Future<Output = ()> + Unpin> AsyncWrite for ShutdownAddrStream<F> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let this = Pin::into_inner(self);
+        if let Err(e) = this.check_shutdown(cx) {
+            return Poll::Ready(Err(e));
+        }
+
+        match this {
+            Self::Alive { inner, .. } => Pin::new(inner).poll_write(cx, buf),
```

Comment:
If `let Poll::Ready(Ok(..)) = self.inner.poll_shutdown()`, I think this will try to write to a closed stream and error

Reply:
That's the intention. We don't have choice other than to error out in this case right?

Reply:
Yeah, I think it's alright to error out, but we should be careful to catch the error upstream so we don't, for example, crash the `live` server

Reply:
It's handled by `hyper`/`tonic` or whatever framework code. As service implementors we won't see this error.